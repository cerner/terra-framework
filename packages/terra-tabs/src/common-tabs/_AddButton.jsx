import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import IconAdd from 'terra-icon/lib/icon/IconAdd';
import Button from 'terra-button';
import { injectIntl } from 'react-intl';
import { KEY_SPACE, KEY_RETURN } from 'keycode-js';
import {
  handleArrows,
} from './_TabUtils';

import styles from './Tab.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The label to set on the add button.
   */
  addAriaLabel: PropTypes.string.isRequired,
  /**
   * id value to use for navigation.
   */
  id: PropTypes.string.isRequired,
  /**
   * Index value to use for navigation.
   */
  index: PropTypes.number.isRequired,
  /**
   *  Callback function triggered when add button is clicked.
  */
  onAddClick: PropTypes.func,
  /**
   * Ref callback for button.
   */
  refCallback: PropTypes.func,
  /**
   * Array of id strings,
   */
  tabIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  /**
   * @private
   * intl object programmatically imported through injectIntl from react-intl.
   */
  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
};

const AddButton = ({
  addAriaLabel,
  onAddClick,
  refCallback,
  index,
  id,
  tabIds,
  intl,
}) => {
  const buttonsuggestion = intl.formatMessage({ id: 'Terra.tabs.addbutton.focus' });
  let ariaLabelAddTabWithSuggestion = `${buttonsuggestion}`;

  if (addAriaLabel) {
    ariaLabelAddTabWithSuggestion = `${addAriaLabel} ${buttonsuggestion}`;
  }
  const AddTabIcon = <IconAdd a11yLabel={ariaLabelAddTabWithSuggestion} />;

  const keyDown = (event, indexTab, ids) => {
    if (event.nativeEvent.keyCode === KEY_RETURN || event.nativeEvent.keyCode === KEY_SPACE) {
      event.preventDefault();
      event.stopPropagation();
      onAddClick();
    } else {
      handleArrows(event, indexTab, ids);
    }
  };

  return (
    <Button
      icon={AddTabIcon}
      isIconOnly
      id={id}
      tabindex="-1"
      index={index}
      text={addAriaLabel}
      variant="de-emphasis"
      data-terra-tabs-show-focus-styles
      className={cx('button')}
      onKeyDown={event => keyDown(event, index, tabIds)}
      ref={refCallback}
      onClick={onAddClick}
      data-terra-tab-add-btn
    />
  );
};

AddButton.propTypes = propTypes;

export default injectIntl(AddButton);
