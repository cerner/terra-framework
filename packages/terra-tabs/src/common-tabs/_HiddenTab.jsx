import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import IconCheckmark from 'terra-icon/lib/icon/IconCheckmark';
import { KEY_SPACE, KEY_RETURN } from 'keycode-js';
import IconClose from 'terra-icon/lib/icon/IconClose';
import { injectIntl } from 'react-intl';
import {
  enableFocusStyles,
  disableFocusStyles,
  handleArrows,
} from './_TabUtils';

import styles from './HiddenTab.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The id of the tab to be used in mapping.
   */
  id: PropTypes.string.isRequired,
  /**
   * The id of the tab pane element associated to this tab.
   */
  associatedPanelId: PropTypes.string.isRequired,
  /**
   * Index value to use for navigation.
   */
  index: PropTypes.number.isRequired,
  /**
   * Indicates if the tab is currently selected.
   */
  isSelected: PropTypes.bool,
  /**
   * Text to be displayed on the tab.
   */
  label: PropTypes.string.isRequired,
  /**
   * Identifer for the Tab to be returned with onSelect.
   */
  itemKey: PropTypes.string.isRequired,
  /**
   * Object to be returned in the onSelect.
   */
  metaData: PropTypes.object,
  /**
   * Callback function triggering on selection. onSelect(itemKey, metaData)
   */
  onSelect: PropTypes.func.isRequired,
  /**
   * Array of id strings,
   */
  tabIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  /**
   * @private
   * The function callback when an event occurs.
   */
  onBlur: PropTypes.func.isRequired,
  /**
   * @private
   * The function callback when an event occurs.
   */
  onFocus: PropTypes.func.isRequired,
  /**
   * @private
   * Callback function when selection has changed.
   * Parameters: 1. Event 2. Selected pane's key
   */
  onChange: PropTypes.func,
  /**
   * Icon to be displayed on the tab.
   */
  icon: PropTypes.element,
  /**
   * If enabled, this prop will show the icon on the tab and also in the menu if pane is collapsed.
   */
  showIcon: PropTypes.bool,
  /**
   * If enabled, this prop will show the add icon in the tab dropdown.
   */
  showAddButton: PropTypes.bool,

  onClosingTab: PropTypes.func,
  isClosable: PropTypes.bool,
  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
  /**
   * Indicates if the pane should be disabled.
   */
  isDisabled: PropTypes.bool,
};

const defaultProps = {
  isSelected: false,
  showIcon: false,
  showAddButton: false,
  isDisabled: false,
};

const HiddenTab = ({
  id,
  associatedPanelId,
  index,
  isSelected,
  label,
  itemKey,
  metaData,
  onBlur,
  onFocus,
  onSelect,
  tabIds,
  onChange,
  icon,
  showIcon,
  showAddButton,
  onClosingTab,
  isClosable,
  intl,
  isDisabled,
}) => {
  const attributes = {};
  const theme = React.useContext(ThemeContext);
  const hiddenClassNames = cx(
    'hidden',
    { 'is-active': isSelected },
    { 'is-disabled': isDisabled },
    theme.className,
  );
  const tabDeleteLabel = `${intl.formatMessage({ id: 'Terra.tabs.hint.removable' })}`;

  const handleOnSelect = (event) => {
    event.preventDefault();
    event.stopPropagation();

    enableFocusStyles(event);
    onSelect(itemKey, metaData);
    if (onChange) {
      onChange(event, itemKey);
    }
  };

  const onKeyDown = (event) => {
    if (event.nativeEvent.keyCode === KEY_RETURN || event.nativeEvent.keyCode === KEY_SPACE) {
      handleOnSelect(event);
      if (onChange) {
        onChange(event, itemKey);
      }
    } else {
      handleArrows(event, index, tabIds);
    }
  };
  function onCloseClick(event) {
    event.stopPropagation();
    // setIsClosed(true);
    onClosingTab(itemKey, metaData);
  }

  function onClick(e) {
    if (!isDisabled) {
      e.preventDefault();
      e.stopPropagation();
      handleOnSelect(e);
    }
  }

  attributes.tabIndex = isSelected ? 0 : -1;
  attributes.onClick = onClick;
  attributes.onKeyDown = onKeyDown;
  attributes.onBlur = e => { enableFocusStyles(e); onBlur(e); };
  attributes.onFocus = onFocus;
  attributes.onMouseDown = disableFocusStyles;
  attributes['data-focus-styles-enabled'] = true;
  attributes['aria-selected'] = isSelected;

  return (
    <div
      {...attributes}
      id={id}
      aria-controls={associatedPanelId}
      role={showAddButton ? 'button' : 'tab'}
      className={hiddenClassNames}
      aria-disabled={isDisabled}
    >
      <div className={cx('checkbox')}>{isSelected ? <IconCheckmark /> : null}</div>
      {showIcon && icon}
      <div className={cx('label', { 'with-icon': showIcon })}>{label}</div>
      {isClosable && (
      <button
        className={cx('pill-remove-button')}
        type="button"
        aria-label={tabDeleteLabel}
        onClick={onCloseClick}
      >
        <IconClose a11yLabel="Closed CLICKED" />
      </button>
      )}
    </div>
  );
};

HiddenTab.propTypes = propTypes;
HiddenTab.defaultProps = defaultProps;

export default injectIntl(HiddenTab);
