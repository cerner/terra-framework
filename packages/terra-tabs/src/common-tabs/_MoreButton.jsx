import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import IconCaretDown from 'terra-icon/lib/icon/IconCaretDown';
import { FormattedMessage, injectIntl } from 'react-intl';
import {
  handleMoreButtonArrows,
} from './_TabUtils';

import styles from './Tab.module.scss';
import terraStyles from './TerraTabs.module.scss';

const cx = classNames.bind(styles);
const cy = classNames.bind(terraStyles);

const propTypes = {
  /**
   * The index of the first hidden item in the dropdown.
   */
  hiddenIndex: PropTypes.number.isRequired,
  /**
   * Whether or not the an active item is present.
   */
  isActive: PropTypes.bool,
  /**
   * Whether or not the associated dropdown is open.
   */
  isOpen: PropTypes.bool,
  /**
   * Blur callback function.
   */
  onBlur: PropTypes.func,
  /**
   * Selection callback function.
   */
  onSelect: PropTypes.func,
  /**
   * Ref callback for button.
   */
  refCallback: PropTypes.func,
  /**
   * Array of id strings,
   */
  tabIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  /**
   * The z-index style to apply to the button based upon order and state.
   */
  zIndex: PropTypes.number,
  /**
   *  @private
   * The style to be applied to the tabs
   */
  variant: PropTypes.oneOf(['workspace', 'framework']),
};

const setFocus = event => {
  event.currentTarget.setAttribute('tabindex', '-1');
  event.currentTarget.focus();
};

const removeFocus = event => {
  event.currentTarget.removeAttribute('tabindex');
};

const MoreButton = ({
  hiddenIndex,
  isActive,
  isOpen,
  onBlur,
  onSelect,
  refCallback,
  tabIds,
  zIndex,
  variant,
}) => {
  const theme = React.useContext(ThemeContext);

  const handleOnKeyDown = event => handleMoreButtonArrows(event, hiddenIndex, tabIds);

  const handleOnMouseDown = event => setFocus(event);

  const handleOnSelect = event => {
    setFocus(event);
    if (onSelect) {
      onSelect(event);
    }
  };

  const handleOnBlur = event => {
    removeFocus(event);
    if (onBlur) {
      onBlur(event);
    }
  };
  const buttonClass = variant === 'framework' ? cy('tab-menu', theme.className) : cx('tab-menu', { 'is-active': isOpen || isActive }, { 'is-open': isOpen }, theme.className);

  /* eslint-disable react/forbid-dom-props */
  return (
    <div
      aria-hidden
      role="button"
      ref={refCallback}
      onClick={handleOnSelect}
      onKeyDown={handleOnKeyDown}
      onBlur={handleOnBlur}
      onMouseDown={handleOnMouseDown}
      className={buttonClass}
      style={{ zIndex: isOpen ? '100' : zIndex }}
      data-testid="workspace-tabs-more-button"
    >
      {variant === 'framework'
        ? (
          <FormattedMessage id="Terra.tabs.more">
            {menuToggleText => (
              <span>{menuToggleText}</span>
            )}
          </FormattedMessage>
        ) : null}
      <div className={cx('inner')}>
        <div className={cx('icon')}>
          <IconCaretDown />
        </div>
      </div>
    </div>
  );
  /* eslint-disable react/forbid-dom-props */
};

MoreButton.propTypes = propTypes;

export default injectIntl(MoreButton);
