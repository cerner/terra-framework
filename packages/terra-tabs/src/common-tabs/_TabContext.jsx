import { createContext } from 'react';

const TabContext = createContext();

export default TabContext;
