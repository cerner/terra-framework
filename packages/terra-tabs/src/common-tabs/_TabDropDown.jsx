import React, { useCallback, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import onClickOutside from 'react-onclickoutside';
import { KEY_ESCAPE } from 'keycode-js';

import styles from './TabDropDown.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Currently active HiddenTabs be displayed.
   */
  children: PropTypes.node,
  /**
   * The id of the tab to be used in mapping.
   */
  isOpen: PropTypes.bool,
  /**
   * The function returning the html node.
   */
  refCallback: PropTypes.func.isRequired,
  /**
   * @private
   * The function callback when a click outside event occurs.
   */
  onRequestClose: PropTypes.func,
  /**
   * @private
   * The function callback for disable.
   */
  disableOnClickOutside: PropTypes.func,
  /**
   * @private
   * The function callback for enable.
   */
  enableOnClickOutside: PropTypes.func,
};

const TabDropDown = ({
  isOpen,
  children,
  onRequestClose,
  refCallback,
  disableOnClickOutside,
  enableOnClickOutside,
}) => {
  const dropDownRef = useRef();
  const handleKeyDown = useCallback(event => {
    if (event.keyCode === KEY_ESCAPE && onRequestClose) {
      onRequestClose(event);
    }
  }, [onRequestClose]);

  useEffect(() => {
    if (isOpen) {
      enableOnClickOutside();
      document.addEventListener('keydown', handleKeyDown);
    } else {
      disableOnClickOutside();
      document.removeEventListener('keydown', handleKeyDown);
      dropDownRef.current.scrollTop = 0;
    }

    return (() => {
      disableOnClickOutside();
      document.removeEventListener('keydown', handleKeyDown);
    });
  }, [isOpen, enableOnClickOutside, disableOnClickOutside, handleKeyDown]);

  TabDropDown.handleClickOutside = event => onRequestClose(event);

  const theme = React.useContext(ThemeContext);
  const dropDownClassNames = cx(
    'drop-down',
    { 'is-open': isOpen },
    theme.className,
  );

  return (
    <div
      ref={node => { dropDownRef.current = node; refCallback(node); }}
      role="none"
      className={dropDownClassNames}
      onMouseDown={e => { e.preventDefault(); }}
    >
      {children}
    </div>
  );
};

TabDropDown.propTypes = propTypes;

const clickOutsideConfig = {
  handleClickOutside: () => TabDropDown.handleClickOutside,
};

export default onClickOutside(TabDropDown, clickOutsideConfig);
