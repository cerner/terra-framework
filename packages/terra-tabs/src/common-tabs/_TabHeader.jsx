import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import VisuallyHiddenText from 'terra-visually-hidden-text';

import styles from './TabHeader.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The child content to display within the title area
   */
  title: PropTypes.string,
};

const TabHeader = ({
  title,
}) => {
  const theme = React.useContext(ThemeContext);

  // tabIndex of -1 ensure a focus location for the dropdown to navigate to with a screen reader.
  return (
    <div className={cx('header', theme.className)}>
      <VisuallyHiddenText aria-level="2" role="heading" tabIndex="-1" text={title} />
      <div className={cx('title')} aria-hidden="true">
        {title}
      </div>
    </div>
  );
};

TabHeader.propTypes = propTypes;

export default TabHeader;
