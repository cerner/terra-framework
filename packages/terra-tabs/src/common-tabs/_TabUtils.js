import {
  KEY_LEFT,
  KEY_RIGHT,
  KEY_UP,
  KEY_DOWN,
  KEY_HOME,
  KEY_END,
  KEY_RETURN,
  KEY_SPACE,
} from 'keycode-js';

/**
 * Returns a function that will execute the provided function upon detection of a KEY_RETURN or KEY_SPACE keydown event.
 * @param {Function} onSelect The function to be executed after event detection.
 */
const generateKeyDownSelection = onSelect => (
  onSelect ? (event) => {
    if (event.nativeEvent.keyCode === KEY_RETURN || event.nativeEvent.keyCode === KEY_SPACE) {
      event.preventDefault();
      onSelect();
    }
  } : undefined
);

/**
 * Enables focus styles for the target of the given event. Typically used as an onBlur callback on selectable elements.
 */
const enableFocusStyles = (event) => {
  event.currentTarget.setAttribute('data-focus-styles-enabled', 'true');
};

/**
 * Disables focus styles for the target of the given event. Typically used as an onMouseDown callback on selectable elements.
 */
const disableFocusStyles = (event) => {
  event.currentTarget.setAttribute('data-focus-styles-enabled', 'false');
};

/**
 * Sets focus to the 'next' tab element.
 * @param {object} event The triggering event.
 * @param {number} index The index of the current element.
 * @param {array} ids The array of id strings.
 */
const nextFocus = (event, index, ids) => {
  event.preventDefault();
  event.stopPropagation();

  if (index < 0) {
    const element = document.getElementById(ids[0]);
    if (element) {
      element.focus();
    }
    return;
  }

  const newIndex = index + 1 >= ids.length ? 0 : index + 1;
  const element = document.getElementById(ids[newIndex]);
  if (element) {
    element.focus();
  }
};

/**
 * Sets focus to the 'previous' tab element.
 * @param {object} event The triggering event.
 * @param {number} index The index of the current element.
 * @param {array} ids The array of id strings.
 */
const previousFocus = (event, index, ids) => {
  event.preventDefault();
  event.stopPropagation();

  if (index < 0) {
    const element = document.getElementById(ids[ids.length - 1]);
    if (element) {
      element.focus();
    }
    return;
  }

  const newIndex = index - 1 < 0 ? ids.length - 1 : index - 1;
  const element = document.getElementById(ids[newIndex]);
  if (element) {
    element.focus();
  }
};

/**
 * Sets focus to the 'first' tab element.
 * @param {object} event The triggering event.
 * @param {number} index The index of the current element.
 * @param {array} ids The array of id strings.
 */
const firstFocus = (event, index, ids) => {
  event.preventDefault();
  event.stopPropagation();

  if (index !== 0) {
    const element = document.getElementById(ids[0]);
    if (element) {
      element.focus();
    }
  }
};

/**
 * Sets focus to the 'end' tab element.
 * @param {object} event The triggering event.
 * @param {number} index The index of the current element.
 * @param {array} ids The array of id strings.
 */
const lastFocus = (event, index, ids) => {
  event.preventDefault();
  event.stopPropagation();

  if (index !== ids.length - 1) {
    const element = document.getElementById(ids[ids.length - 1]);
    if (element) {
      element.focus();
    }
  }
};

/**
 * The default event handler for navigational arrow keys.
 * @param {object} event The triggering event.
 * @param {number} index The index of the current element.
 * @param {array} ids The array of id strings.
 */
const handleArrows = (event, index, ids) => {
  const isRTL = document.getElementsByTagName('html')[0].getAttribute('dir') === 'rtl';
  const nextKey = !isRTL ? KEY_RIGHT : KEY_LEFT;
  const previousKey = !isRTL ? KEY_LEFT : KEY_RIGHT;
  if (event.nativeEvent.keyCode === nextKey || event.nativeEvent.keyCode === KEY_DOWN) {
    nextFocus(event, index, ids);
  } else if (event.nativeEvent.keyCode === previousKey || event.nativeEvent.keyCode === KEY_UP) {
    previousFocus(event, index, ids);
  } else if (event.nativeEvent.keyCode === KEY_HOME) {
    firstFocus(event, index, ids);
  } else if (event.nativeEvent.keyCode === KEY_END) {
    lastFocus(event, index, ids);
  }
};

/**
 * The more button's event handler for navigational arrow keys.
 * @param {object} event The triggering event.
 * @param {number} hiddenIndex The index of the first hidden index.
 * @param {array} ids The array of id strings.
 */
const handleMoreButtonArrows = (event, hiddenIndex, ids) => {
  const isRTL = document.getElementsByTagName('html')[0].getAttribute('dir') === 'rtl';
  const nextKey = !isRTL ? KEY_RIGHT : KEY_LEFT;
  const previousKey = !isRTL ? KEY_LEFT : KEY_RIGHT;

  if (event.nativeEvent.keyCode === nextKey || event.nativeEvent.keyCode === KEY_DOWN) {
    event.preventDefault();
    event.stopPropagation();

    const element = document.getElementById(ids[hiddenIndex]);
    if (element) {
      element.focus();
    }
  } else if (event.nativeEvent.keyCode === KEY_HOME) {
    firstFocus(event, -1, ids);
  } else if (event.nativeEvent.keyCode === previousKey || event.nativeEvent.keyCode === KEY_UP || event.nativeEvent.keyCode === KEY_END) {
    lastFocus(event, -1, ids);
  }
};

export {
  enableFocusStyles,
  disableFocusStyles,
  generateKeyDownSelection,
  handleArrows,
  handleMoreButtonArrows,
  nextFocus,
  previousFocus,
  firstFocus,
  lastFocus,
};
