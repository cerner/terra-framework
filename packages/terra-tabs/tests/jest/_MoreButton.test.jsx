/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { IntlProvider } from 'react-intl';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';

import MoreButton from '../../src/common-tabs/_MoreButton';

import * as TabUtils from '../../src/common-tabs/_TabUtils';

jest.mock('../../src/common-tabs/_TabUtils', () => ({
  handleMoreButtonArrows: jest.fn(),
}));

describe('MoreButton', () => {
  test('should render a more button with provided props', () => {
    const testTabIds = ['tab-1', 'tab-2', 'tab-3'];
    const wrapper = mountWithIntl(
      <IntlProvider>
        <MoreButton
          hiddenIndex={1}
          tabIds={testTabIds}
          zIndex={0}
        />
      </IntlProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render a more button with provided props and selection handling', () => {
    const mockOnSelect = jest.fn();
    const mockOnBlur = jest.fn();
    const testTabIds = ['tab-1', 'tab-2', 'tab-3'];

    render((
      <IntlProvider>
        <MoreButton
          hiddenIndex={1}
          onBlur={mockOnBlur}
          onSelect={mockOnSelect}
          tabIds={testTabIds}
          zIndex={0}
        />
      </IntlProvider>
    ));

    const buttonElement = screen.getByRole('button', { hidden: true });
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveStyle({ 'z-index': '0' });

    userEvent.click(buttonElement);

    expect(mockOnSelect).toHaveBeenCalled();
    mockOnSelect.mockClear();

    buttonElement.focus();
    userEvent.type(buttonElement, '{arrowright}', { skipClick: true });
    expect(mockOnSelect).not.toHaveBeenCalledWith();
    expect(TabUtils.handleMoreButtonArrows).toHaveBeenCalledWith(expect.anything(), 1, testTabIds);
    mockOnSelect.mockClear();

    buttonElement.blur();
    expect(mockOnBlur).toHaveBeenCalled();
  });

  test('should render a more button when active and open', () => {
    const mockOnSelect = jest.fn();
    const mockOnBlur = jest.fn();
    const testTabIds = ['tab-1', 'tab-2', 'tab-3'];

    render((
      <IntlProvider>
        <MoreButton
          hiddenIndex={1}
          onBlur={mockOnBlur}
          onSelect={mockOnSelect}
          tabIds={testTabIds}
          zIndex={0}
          isActive
          isOpen
        />
      </IntlProvider>
    ));

    const buttonElement = screen.getByRole('button', { hidden: true });
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveClass('is-active');
    expect(buttonElement).toHaveStyle({ 'z-index': '100' });

    userEvent.click(buttonElement);

    expect(mockOnSelect).toHaveBeenCalled();
    mockOnSelect.mockClear();

    buttonElement.focus();
    userEvent.type(buttonElement, '{arrowright}', { skipClick: true });
    expect(mockOnSelect).not.toHaveBeenCalledWith();
    expect(TabUtils.handleMoreButtonArrows).toHaveBeenCalledWith(expect.anything(), 1, testTabIds);
    mockOnSelect.mockClear();

    buttonElement.blur();
    expect(buttonElement).not.toHaveAttribute('tabindex');
    expect(mockOnBlur).toHaveBeenCalled();
  });

  test('should not blow up when onBlur is not provided', () => {
    const mockOnSelect = jest.fn();
    const testTabIds = ['tab-1', 'tab-2', 'tab-3'];

    render((
      <IntlProvider>
        <MoreButton
          hiddenIndex={1}
          onSelect={mockOnSelect}
          tabIds={testTabIds}
          zIndex={0}
        />
      </IntlProvider>
    ));

    const buttonElement = screen.getByRole('button', { hidden: true });
    expect(buttonElement).toBeInTheDocument();

    buttonElement.focus();
    buttonElement.blur();

    expect(buttonElement).not.toHaveAttribute('tabindex');
  });
});
