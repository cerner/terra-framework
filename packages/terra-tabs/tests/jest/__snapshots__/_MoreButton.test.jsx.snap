// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreButton should render a more button with provided props and selection handling 1`] = `
<IntlProvider
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
>
  <InjectIntl(MoreButton)
    hiddenIndex={1}
    onBlur={
      [MockFunction] {
        "calls": Array [
          Array [
            Class {
              "_dispatchInstances": null,
              "_dispatchListeners": null,
              "_targetInst": null,
              "bubbles": null,
              "cancelable": null,
              "currentTarget": [Function],
              "defaultPrevented": null,
              "detail": null,
              "dispatchConfig": null,
              "eventPhase": null,
              "isDefaultPrevented": [Function],
              "isPropagationStopped": [Function],
              "isTrusted": null,
              "nativeEvent": null,
              "relatedTarget": null,
              "target": null,
              "timeStamp": [Function],
              "type": null,
              "view": null,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    onSelect={[MockFunction]}
    tabIds={
      Array [
        "tab-1",
        "tab-2",
        "tab-3",
      ]
    }
    zIndex={0}
  >
    <MoreButton
      hiddenIndex={1}
      intl={
        Object {
          "defaultFormats": Object {},
          "defaultLocale": "en",
          "formatDate": [Function],
          "formatHTMLMessage": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatPlural": [Function],
          "formatRelative": [Function],
          "formatTime": [Function],
          "formats": Object {},
          "formatters": Object {
            "getDateTimeFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralFormat": [Function],
            "getRelativeFormat": [Function],
          },
          "locale": "en",
          "messages": null,
          "now": [Function],
          "onError": [Function],
          "textComponent": "span",
          "timeZone": null,
        }
      }
      onBlur={
        [MockFunction] {
          "calls": Array [
            Array [
              Class {
                "_dispatchInstances": null,
                "_dispatchListeners": null,
                "_targetInst": null,
                "bubbles": null,
                "cancelable": null,
                "currentTarget": [Function],
                "defaultPrevented": null,
                "detail": null,
                "dispatchConfig": null,
                "eventPhase": null,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": null,
                "nativeEvent": null,
                "relatedTarget": null,
                "target": null,
                "timeStamp": [Function],
                "type": null,
                "view": null,
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      onSelect={[MockFunction]}
      tabIds={
        Array [
          "tab-1",
          "tab-2",
          "tab-3",
        ]
      }
      zIndex={0}
    >
      <div
        aria-hidden={true}
        className="tab-menu"
        data-testid="workspace-tabs-more-button"
        onBlur={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        style={
          Object {
            "zIndex": 0,
          }
        }
      >
        <div
          className="inner"
        >
          <div
            className="icon"
          >
            <IconCaretDown
              className=""
              isBidi={true}
              viewBox="0 0 48 48"
              xmlns="http://www.w3.org/2000/svg"
            >
              <IconBase
                className=""
                focusable="false"
                height="1em"
                isBidi={true}
                isSpin={false}
                viewBox="0 0 48 48"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <svg
                  className="tui-Icon icon is-bidi"
                  focusable="false"
                  height="1em"
                  role="presentation"
                  viewBox="0 0 48 48"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M48 12L24 36 0 12h48z"
                  />
                </svg>
              </IconBase>
            </IconCaretDown>
          </div>
        </div>
      </div>
    </MoreButton>
  </InjectIntl(MoreButton)>
</IntlProvider>
`;
